File: sdks/py_sdk/sw4rm/llm/rate_limiter.py
"""Token bucket rate limiter for LLM API requests.

Provides proactive rate limiting to avoid 429 errors. All LLM clients
in the process share a global singleton bucket.

Configuration via environment variables:
    LLM_RATE_LIMIT_ENABLED: "1" (default) or "0"
    LLM_RATE_LIMIT_TOKENS_PER_MIN: 250000 (default, matches Groq free tier)
    LLM_RATE_LIMIT_ADAPTIVE: "1" (default) — reduce budget on 429, recover on success
"""

import asyncio
import logging
import os
import time
from dataclasses import dataclass, field
from typing import Optional

logger = logging.getLogger(__name__)


@dataclass
class RateLimiterConfig:
    """Configuration for the token bucket rate limiter."""

    tokens_per_minute: int = field(
        default_factory=lambda: int(os.getenv("LLM_RATE_LIMIT_TOKENS_PER_MIN", "250000"))
    )
    burst_allowance: float = 1.0
    min_tokens_per_request: int = 100
    max_wait_seconds: float = 120.0
    enabled: bool = field(
        default_factory=lambda: os.getenv("LLM_RATE_LIMIT_ENABLED", "1").lower()
        not in ("0", "false", "no")
    )
    adaptive_enabled: bool = field(
        default_factory=lambda: os.getenv("LLM_RATE_LIMIT_ADAPTIVE", "1").lower()
        not in ("0", "false", "no")
    )
    reduction_factor: float = field(
        default_factory=lambda: float(os.getenv("LLM_RATE_LIMIT_REDUCTION_FACTOR", "0.7"))
    )
    recovery_factor: float = field(
        default_factory=lambda: float(os.getenv("LLM_RATE_LIMIT_RECOVERY_FACTOR", "1.1"))
    )
    cooldown_seconds: float = field(
        default_factory=lambda: float(os.getenv("LLM_RATE_LIMIT_COOLDOWN_SECONDS", "30"))
    )
    successes_for_recovery: int = field(
        default_factory=lambda: int(os.getenv("LLM_RATE_LIMIT_RECOVERY_SUCCESS_THRESHOLD", "20"))
    )


class TokenBucket:
    """Token bucket rate limiter with adaptive throttling.

    Refills tokens at a steady rate. When a 429 is reported via
    record_rate_limit(), the budget is reduced. After enough successes
    and a cooldown period, the budget recovers.
    """

    def __init__(self, config: Optional[RateLimiterConfig] = None):
        self.config = config or RateLimiterConfig()
        self._base_tpm = float(self.config.tokens_per_minute)
        self._current_tpm = float(self.config.tokens_per_minute)
        self._min_tpm = max(1000.0, self._base_tpm * 0.25)
        self._tokens = self._current_tpm  # start full
        self._last_refill = time.monotonic()
        self._lock = asyncio.Lock()
        self._last_rate_limit_time: Optional[float] = None
        self._successes_since_limit = 0

    def _refill(self) -> None:
        now = time.monotonic()
        elapsed = now - self._last_refill
        refill = elapsed * (self._current_tpm / 60.0)
        self._tokens = min(
            self._tokens + refill,
            self._current_tpm * self.config.burst_allowance,
        )
        self._last_refill = now
        self._maybe_recover(now)

    def _maybe_recover(self, now: float) -> None:
        if not self.config.adaptive_enabled:
            return
        if self._current_tpm >= self._base_tpm:
            return
        if self._last_rate_limit_time is None:
            return
        if (now - self._last_rate_limit_time) < self.config.cooldown_seconds:
            return
        if self._successes_since_limit < self.config.successes_for_recovery:
            return
        new_limit = min(self._base_tpm, self._current_tpm * self.config.recovery_factor)
        if new_limit > self._current_tpm:
            self._current_tpm = new_limit
            self._successes_since_limit = 0
            logger.info("Rate limiter recovered to %.0f tokens/min", self._current_tpm)

    async def acquire(self, estimated_tokens: int) -> float:
        """Acquire tokens, waiting if necessary.

        Returns time spent waiting (0.0 if immediate).
        """
        if not self.config.enabled:
            return 0.0

        estimated_tokens = max(estimated_tokens, self.config.min_tokens_per_request)
        wait_start = time.monotonic()

        while True:
            async with self._lock:
                self._refill()
                if self._tokens >= estimated_tokens:
                    self._tokens -= estimated_tokens
                    return time.monotonic() - wait_start

            elapsed = time.monotonic() - wait_start
            if elapsed >= self.config.max_wait_seconds:
                raise asyncio.TimeoutError(
                    f"Rate limiter: waited {elapsed:.1f}s for {estimated_tokens} tokens"
                )
            await asyncio.sleep(0.25)

    async def record_rate_limit(self) -> None:
        """Record a 429 event — adaptively reduce budget."""
        if not (self.config.enabled and self.config.adaptive_enabled):
            return
        async with self._lock:
            self._last_rate_limit_time = time.monotonic()
            self._successes_since_limit = 0
            new_limit = max(self._min_tpm, self._current_tpm * self.config.reduction_factor)
            if new_limit < self._current_tpm:
                self._current_tpm = new_limit
                self._tokens = min(self._tokens, self._current_tpm)
                logger.warning("Rate limiter reduced to %.0f tokens/min", self._current_tpm)

    async def record_success(self) -> None:
        """Record a successful request for adaptive recovery."""
        if not (self.config.enabled and self.config.adaptive_enabled):
            return
        async with self._lock:
            self._successes_since_limit += 1

    @property
    def available_tokens(self) -> float:
        return self._tokens

    @property
    def current_tokens_per_minute(self) -> float:
        return self._current_tpm


# Global singleton
_global_bucket: Optional[TokenBucket] = None


def get_global_rate_limiter(config: Optional[RateLimiterConfig] = None) -> TokenBucket:
    """Get or create the global rate limiter singleton."""
    global _global_bucket
    if _global_bucket is None:
        _global_bucket = TokenBucket(config)
    return _global_bucket


def reset_global_rate_limiter() -> None:
    """Reset the global rate limiter (for testing)."""
    global _global_bucket
    _global_bucket = None

File: sdks/js_sdk/src/llm/rateLimiter.ts
/**
 * Token bucket rate limiter for LLM API requests.
 *
 * Provides proactive rate limiting to avoid 429 errors. All LLM clients
 * in the process share a global singleton bucket.
 *
 * Configuration via environment variables:
 *   LLM_RATE_LIMIT_ENABLED:          "1" (default) or "0"
 *   LLM_RATE_LIMIT_TOKENS_PER_MIN:   250000 (default, matches Groq free tier)
 *   LLM_RATE_LIMIT_ADAPTIVE:         "1" (default) -- reduce budget on 429, recover on success
 *   LLM_RATE_LIMIT_REDUCTION_FACTOR: 0.7
 *   LLM_RATE_LIMIT_RECOVERY_FACTOR:  1.1
 *   LLM_RATE_LIMIT_COOLDOWN_SECONDS: 30
 *   LLM_RATE_LIMIT_RECOVERY_SUCCESS_THRESHOLD: 20
 *
 * @module llm/rateLimiter
 */

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

/** Configuration for the token bucket rate limiter. */
export interface RateLimiterConfig {
  /** Token budget per minute. */
  tokensPerMinute: number;
  /** Burst allowance multiplier (1.0 = no burst above budget). */
  burstAllowance: number;
  /** Minimum tokens assumed per request. */
  minTokensPerRequest: number;
  /** Maximum seconds to wait before throwing a timeout. */
  maxWaitSeconds: number;
  /** Whether the rate limiter is enabled at all. */
  enabled: boolean;
  /** Whether adaptive throttling is enabled. */
  adaptiveEnabled: boolean;
  /** Factor to multiply budget by after a 429 event. */
  reductionFactor: number;
  /** Factor to multiply budget by during recovery. */
  recoveryFactor: number;
  /** Seconds to wait after last 429 before allowing recovery. */
  cooldownSeconds: number;
  /** Number of consecutive successes required before recovery. */
  successesForRecovery: number;
}

/** Parse an env var as a boolean, treating "0", "false", "no" as false. */
function envBool(name: string, defaultVal: string): boolean {
  const raw = (process.env[name] ?? defaultVal).toLowerCase();
  return !['0', 'false', 'no'].includes(raw);
}

/** Build a RateLimiterConfig from environment variables and optional overrides. */
export function buildRateLimiterConfig(
  overrides?: Partial<RateLimiterConfig>,
): RateLimiterConfig {
  return {
    tokensPerMinute:
      overrides?.tokensPerMinute ??
      parseInt(process.env['LLM_RATE_LIMIT_TOKENS_PER_MIN'] ?? '250000', 10),
    burstAllowance: overrides?.burstAllowance ?? 1.0,
    minTokensPerRequest: overrides?.minTokensPerRequest ?? 100,
    maxWaitSeconds: overrides?.maxWaitSeconds ?? 120.0,
    enabled: overrides?.enabled ?? envBool('LLM_RATE_LIMIT_ENABLED', '1'),
    adaptiveEnabled:
      overrides?.adaptiveEnabled ??
      envBool('LLM_RATE_LIMIT_ADAPTIVE', '1'),
    reductionFactor:
      overrides?.reductionFactor ??
      parseFloat(process.env['LLM_RATE_LIMIT_REDUCTION_FACTOR'] ?? '0.7'),
    recoveryFactor:
      overrides?.recoveryFactor ??
      parseFloat(process.env['LLM_RATE_LIMIT_RECOVERY_FACTOR'] ?? '1.1'),
    cooldownSeconds:
      overrides?.cooldownSeconds ??
      parseFloat(process.env['LLM_RATE_LIMIT_COOLDOWN_SECONDS'] ?? '30'),
    successesForRecovery:
      overrides?.successesForRecovery ??
      parseInt(
        process.env['LLM_RATE_LIMIT_RECOVERY_SUCCESS_THRESHOLD'] ?? '20',
        10,
      ),
  };
}

// ---------------------------------------------------------------------------
// TokenBucket
// ---------------------------------------------------------------------------

/**
 * Token bucket rate limiter with adaptive throttling.
 *
 * Refills tokens at a steady rate. When a 429 is reported via
 * {@link recordRateLimit}, the budget is reduced. After enough successes
 * and a cooldown period, the budget recovers.
 */
export class TokenBucket {
  private readonly config: RateLimiterConfig;
  private readonly baseTpm: number;
  private currentTpm: number;
  private readonly minTpm: number;
  private tokens: number;
  private lastRefill: number;
  private lastRateLimitTime: number | null = null;
  private successesSinceLimit = 0;

  constructor(config?: Partial<RateLimiterConfig>) {
    this.config = buildRateLimiterConfig(config);
    this.baseTpm = this.config.tokensPerMinute;
    this.currentTpm = this.config.tokensPerMinute;
    this.minTpm = Math.max(1000, this.baseTpm * 0.25);
    this.tokens = this.currentTpm; // start full
    this.lastRefill = performance.now();
  }

  // -- Internal helpers ----------------------------------------------------

  private refill(): void {
    const now = performance.now();
    const elapsedSeconds = (now - this.lastRefill) / 1000;
    const refill = elapsedSeconds * (this.currentTpm / 60);
    this.tokens = Math.min(
      this.tokens + refill,
      this.currentTpm * this.config.burstAllowance,
    );
    this.lastRefill = now;
    this.maybeRecover(now);
  }

  private maybeRecover(now: number): void {
    if (!this.config.adaptiveEnabled) return;
    if (this.currentTpm >= this.baseTpm) return;
    if (this.lastRateLimitTime === null) return;
    if ((now - this.lastRateLimitTime) / 1000 < this.config.cooldownSeconds) return;
    if (this.successesSinceLimit < this.config.successesForRecovery) return;

    const newLimit = Math.min(
      this.baseTpm,
      this.currentTpm * this.config.recoveryFactor,
    );
    if (newLimit > this.currentTpm) {
      this.currentTpm = newLimit;
      this.successesSinceLimit = 0;
    }
  }

  // -- Public API ----------------------------------------------------------

  /**
   * Acquire tokens, waiting if necessary.
   *
   * @param estimatedTokens - Estimated token count for the request.
   * @returns Time spent waiting in milliseconds (0 if immediate).
   * @throws Error if waiting exceeds {@link RateLimiterConfig.maxWaitSeconds}.
   */
  async acquire(estimatedTokens: number): Promise<number> {
    if (!this.config.enabled) return 0;

    estimatedTokens = Math.max(estimatedTokens, this.config.minTokensPerRequest);
    const waitStart = performance.now();

    for (;;) {
      this.refill();
      if (this.tokens >= estimatedTokens) {
        this.tokens -= estimatedTokens;
        return performance.now() - waitStart;
      }

      const elapsed = (performance.now() - waitStart) / 1000;
      if (elapsed >= this.config.maxWaitSeconds) {
        throw new Error(
          `Rate limiter: waited ${elapsed.toFixed(1)}s for ${estimatedTokens} tokens`,
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 250));
    }
  }

  /**
   * Record a 429 event -- adaptively reduce budget.
   *
   * Call this when the upstream API returns HTTP 429 or an equivalent
   * rate-limit error.
   */
  recordRateLimit(): void {
    if (!(this.config.enabled && this.config.adaptiveEnabled)) return;

    this.lastRateLimitTime = performance.now();
    this.successesSinceLimit = 0;
    const newLimit = Math.max(
      this.minTpm,
      this.currentTpm * this.config.reductionFactor,
    );
    if (newLimit < this.currentTpm) {
      this.currentTpm = newLimit;
      this.tokens = Math.min(this.tokens, this.currentTpm);
    }
  }

  /**
   * Record a successful request for adaptive recovery.
   *
   * Call this after each successful API response.
   */
  recordSuccess(): void {
    if (!(this.config.enabled && this.config.adaptiveEnabled)) return;
    this.successesSinceLimit += 1;
  }

  /** Current available token count. */
  get availableTokens(): number {
    return this.tokens;
  }

  /** Current tokens-per-minute budget (may be reduced after 429). */
  get currentTokensPerMinute(): number {
    return this.currentTpm;
  }
}

// ---------------------------------------------------------------------------
// Global singleton
// ---------------------------------------------------------------------------

let globalBucket: TokenBucket | null = null;

/**
 * Get or create the global rate limiter singleton.
 *
 * @param config - Optional configuration overrides (only used on first call).
 * @returns The shared TokenBucket instance.
 */
export function getGlobalRateLimiter(
  config?: Partial<RateLimiterConfig>,
): TokenBucket {
  if (globalBucket === null) {
    globalBucket = new TokenBucket(config);
  }
  return globalBucket;
}

/**
 * Reset the global rate limiter (for testing).
 *
 * Clears the singleton so the next call to {@link getGlobalRateLimiter}
 * creates a fresh instance.
 */
export function resetGlobalRateLimiter(): void {
  globalBucket = null;
}
